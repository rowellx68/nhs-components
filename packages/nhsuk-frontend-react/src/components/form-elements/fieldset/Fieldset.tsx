'use client';

import React, { memo, useReducer, useMemo } from 'react';
import clsx from 'clsx';
import {
  FieldsetContextValue,
  FieldsetProvider,
  reducer,
} from './Fieldset.context';
import { Base, BaseProps } from '@/components/core/base/Base';
import { AsElementProps, ElementProps, Size } from '@/types/shared';
import { factory, Factory } from '@/internal/factory/factory';

export type FieldsetProps = ElementProps<'fieldset'>;

type FieldsetFactory = Factory<{
  props: FieldsetProps;
  ref: HTMLFieldSetElement;
  staticComponents: {
    Label: typeof FieldsetLegend;
  };
}>;

const BaseFieldset = factory<FieldsetFactory>(
  ({ className, ...props }, ref) => {
    return (
      <fieldset
        className={clsx('nhsuk-fieldset', className)}
        {...props}
        ref={ref}
      />
    );
  },
);

const MemoisedFieldset = memo(BaseFieldset);

const Fieldset = ({
  disableErrorLine,
  ...props
}: FieldsetProps & { disableErrorLine?: boolean }) => {
  const [state, dispatch] = useReducer(reducer, {
    registered: [],
    errored: [],
  });

  const contextValue: FieldsetContextValue = useMemo(
    () => ({
      isFieldset: true,
      dispatch,
    }),
    [dispatch],
  );

  const withFormElements = state.registered.length > 0;
  const withErrors = state.errored.length > 0;

  const baseProps = withFormElements
    ? {
        as: 'div',
        className: clsx([
          'nhsuk-form-group',
          { 'nhsuk-form-group--error': !disableErrorLine && withErrors },
        ]),
      }
    : {
        as: React.Fragment,
      };

  return (
    <FieldsetProvider value={contextValue}>
      <Base<any> {...baseProps}>
        <MemoisedFieldset {...props} />
      </Base>
    </FieldsetProvider>
  );
};

export type FieldsetLegendProps = (
  | {
      size?: undefined;
      variant?: undefined;
    }
  | {
      size: Exclude<Size, 'xl'>;
      variant?: undefined;
    }
  | {
      variant: 'page-heading';
      size?: Exclude<Size, 'xl'>;
    }
) &
  ElementProps<'legend', 'size' | 'as'> &
  BaseProps;

const FieldsetLegend = ({
  variant,
  size,
  as: component,
  className,
  children,
  ...props
}: FieldsetLegendProps & AsElementProps) => {
  const _component =
    variant === 'page-heading' || size === 'l'
      ? component || 'h1'
      : React.Fragment;

  const baseProps = {
    as: _component,
    ...(_component !== React.Fragment
      ? {
          className:
            variant === 'page-heading' ? 'nhsuk-fieldset__heading' : undefined,
        }
      : {}),
  };

  return (
    <legend
      className={clsx(
        'nhsuk-fieldset__legend',
        {
          'nhsuk-fieldset__legend--l': variant === 'page-heading' && !size,
          [`nhsuk-fieldset__legend--${size}`]: size,
        },
        className,
      )}
      {...props}
    >
      <Base<any> {...baseProps}>{children}</Base>
    </legend>
  );
};

Fieldset.displayName = 'Fieldset';
FieldsetLegend.displayName = 'Fieldset.Legend';

Fieldset.Legend = FieldsetLegend;

export { Fieldset, FieldsetLegend };
