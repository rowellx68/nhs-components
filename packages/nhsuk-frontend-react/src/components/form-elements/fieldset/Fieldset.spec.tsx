import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { Fieldset } from './Fieldset';

it('should render the fieldset component', () => {
  const { container } = render(
    <Fieldset>
      <Fieldset.Legend size="l" variant="page-heading">
        What is your address?
      </Fieldset.Legend>
    </Fieldset>,
  );

  expect(
    container.querySelector('.nhsuk-fieldset__legend--l'),
  ).toHaveTextContent('What is your address?');
  expect(container).toMatchSnapshot();
});
