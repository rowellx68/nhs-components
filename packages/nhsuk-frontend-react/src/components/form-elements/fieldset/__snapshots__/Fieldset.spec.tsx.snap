// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the fieldset component 1`] = `
<div>
  <fieldset
    class="nhsuk-fieldset"
  >
    <legend
      class="nhsuk-legend nhsuk-fieldset__legend--l"
    >
      <h1
        class="nhsuk-fieldset__heading"
      >
        What is your address?
      </h1>
    </legend>
  </fieldset>
</div>
`;
